File: 6_减色处理/python3/DecreaseColor.py
from email.mime import image
import cv2
import numpy as np

def DecreaseColor(srcImage):

    destImage = srcImage.copy()
    
    ind = np.where(destImage < 64)
    destImage[ind] = 32
    
    ind = np.where((destImage >= 64) & (destImage < 128))
    destImage[ind] = 96
    ind = np.where((destImage >= 128) & (destImage < 192))
    destImage[ind] = 160
    ind = np.where((destImage >= 192) & (destImage < 256))
    destImage[ind] = 224

    return destImage


srcImage = cv2.imread("imori.jpg")
cv2.imshow("imori", srcImage)

destImage = DecreaseColor(srcImage)
cv2.imshow("destImage", destImage)


cv2.waitKey(0)
cv2.destroyAllWindows()

File: 6_减色处理/c++/DecreaseColor.cpp

#include <iostream>
#include <opencv2/opencv.hpp>
 
using namespace std;
using namespace cv;
 
Mat DecreaseColor(const Mat& srcImage)
{
    int cols = srcImage.cols;
    int rows = srcImage.rows;
    int channels = srcImage.channels();
    Mat destImage = Mat::zeros(rows, cols, CV_8UC3);

    for (int i = 0; i < rows; ++i)
    {
        for (int j = 0; j < cols; ++j)
        {
            for (int k = 0; k < channels; ++k)
            {
                destImage.ptr<cv::Vec3b>(i)[j][k] = srcImage.ptr<cv::Vec3b>(i)[j][k] / 64 * 64 + 32;
            }
        }
    }

    return destImage;
}

int main()
{
    Mat srcImage = imread("../imori.jpg");
    imshow("Display Image window", srcImage);
    Mat destImage = DecreaseColor(srcImage);
    imshow("destImage", destImage);

    waitKey(0);
    return 0;
}
